src/ZenohUTransport.cpp

```diff
@@ -9,4 +9,384 @@
+
+		// authority_name
+		if (uuri.authority_name().empty()) {
+			zenoh_key << default_authority_name;
+		} else {
+			zenoh_key << uuri.authority_name();
+		}
+		zenoh_key << "/";
+
+		// ue_id
+		if (uuri.ue_id() == WILDCARD_ENTITY_ID) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.ue_id();
+		}
+		zenoh_key << "/";
+
+		// ue_version_major
+		if (uuri.ue_version_major() == WILDCARD_ENTITY_VERSION) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.ue_version_major();
+		}
+		zenoh_key << "/";
+
+		// resource_id
+		if (uuri.resource_id() == WILDCARD_RESOURCE_ID) {
+			zenoh_key << "*";
+		} else {
+			zenoh_key << uuri.resource_id();
+		}
+	};
+
+	zenoh_key << "up";
+	zenoh_key << std::uppercase << std::hex;
+
+	writeUUri(source);
+
+	if (sink.has_value()) {
+		writeUUri(*sink);
+	} else {
+		zenoh_key << "/{}/{}/{}/{}";
+	}
+	return zenoh_key.str();
+}
+
+std::vector<std::pair<std::string, std::string>>
+ZenohUTransport::uattributesToAttachment(const UAttributes& attributes) {
+	std::vector<std::pair<std::string, std::string>> res;
+
+	std::string version(&UATTRIBUTE_VERSION, 1);
+
+	std::string data;
+	attributes.SerializeToString(&data);
+
+	res.push_back(std::make_pair("", version));
+	res.push_back(std::make_pair("", data));
+	return res;
+}
+
+UAttributes ZenohUTransport::attachmentToUAttributes(
+    const AttachmentView& attachment) {
+	std::vector<BytesView> attachment_vec;
+	attachment.iterate(
+	    [&](const BytesView& key, const BytesView& value) -> bool {
+		    attachment_vec.push_back(value);
+		    return true;
+	    });
+
+	if (attachment_vec.size() != 2) {
+		// TODO: error report, exception?
+	}
+
+	if (attachment_vec[0].get_len() == 1) {
+		if (attachment_vec[0].as_string_view()[0] != UATTRIBUTE_VERSION) {
+			// TODO: error report, exception?
+		}
+	};
+	UAttributes res;
+	// TODO: more efficient way?
+	res.ParseFromString(std::string(attachment_vec[1].as_string_view()));
+	return res;
+}
+
+Priority ZenohUTransport::mapZenohPriority(UPriority upriority) {
+	switch (upriority) {
+		case UPriority::UPRIORITY_CS0:
+			return Z_PRIORITY_BACKGROUND;
+		case UPriority::UPRIORITY_CS1:
+			return Z_PRIORITY_DATA_LOW;
+		case UPriority::UPRIORITY_CS2:
+			return Z_PRIORITY_DATA;
+		case UPriority::UPRIORITY_CS3:
+			return Z_PRIORITY_DATA_HIGH;
+		case UPriority::UPRIORITY_CS4:
+			return Z_PRIORITY_INTERACTIVE_LOW;
+		case UPriority::UPRIORITY_CS5:
+			return Z_PRIORITY_INTERACTIVE_HIGH;
+		case UPriority::UPRIORITY_CS6:
+			return Z_PRIORITY_REAL_TIME;
+		case UPriority::UPRIORITY_UNSPECIFIED:
+		default:
+			return Z_PRIORITY_DATA_LOW;
+	}
+}
+
+UMessage ZenohUTransport::sampleToUMessage(const Sample& sample) {
+	UAttributes attributes;
+	if (sample.get_attachment().check()) {
+		attributes = attachmentToUAttributes(sample.get_attachment());
+	}
+	std::string payload(sample.get_payload().as_string_view());
+	UMessage message;
+	message.set_payload(payload);
+	message.set_allocated_attributes(&attributes);
```

Comment:
`set_allocated_attributes` is not safe. Use assignment on the dereferenced mutable attributes:

```suggestion
	message.set_payload(std::move(payload));
	(*message.mutable_attributes()) = std::move(attributes);
```

Reply:
Yes, it was already fixed in my PR